openslides/utils/collection.py

```diff
@@ -117,14 +152,217 @@ def get_full_data(self):
+            self.full_data = cache.get(self.get_cache_key())
+
+        if self.full_data is None:
+            self.full_data = self.get_access_permissions().get_full_data(self.get_instance())
+            self.save_to_cache()
+        return self.full_data
 
     def is_deleted(self):
         """
         Returns Ture if the item is marked as deleted.
         """
         return self.deleted
 
+    def get_cache_key(self):
+        """
+        Returns a string that is used as cache key for a single instance.
+        """
+        return get_single_element_cache_key(self.collection_string, self.id)
+
+    def delete_from_cache(self):
+        """
+        Delets an element from the cache.
+
+        Does nothing if the element is not in the cache.
+        """
+        # Deletes the element from the cache.
+        cache.delete(self.get_cache_key())
+
+        # Delete the id of the instance of the instance list
+        Collection(self.collection_string).delete_id_from_cache(self.id)
+
+    def save_to_cache(self):
+        """
+        Add or update the element to the cache.
+        """
+        # Set the element to the cache.
+        cache.set(self.get_cache_key(), self.get_full_data())
+
+        # Add the id of the element to the collection
+        Collection(self.collection_string).add_id_to_cache(self.id)
+
+
+class Collection:
+    """
+    Represents all elements of one collection.
+    """
+
+    def __init__(self, collection_string):
+        self.collection_string = collection_string
+
+    def get_cache_key(self):
+        """
+        Returns a string that is used as cache key for a collection.
+        """
+        return get_element_list_cache_key(self.collection_string)
+
+    def get_model(self):
+        """
+        Returns the django model that is used for this collection.
+        """
+        return get_model_from_collection_string(self.collection_string)
+
+    def element_generator(self):
+        """
+        Generator that yields all collection_elements of this collection.
+        """
+        # Get all cache keys.
+        ids = self.get_all_ids()
+        cache_keys = [
+            get_single_element_cache_key(self.collection_string, id)
+            for id in ids]
+        cached_full_data_dict = cache.get_many(cache_keys)
+
+        # Get all ids that are missing.
+        missing_ids = set()
+        for cache_key in cache_keys:
+            if cache_key not in cached_full_data_dict:
+                missing_ids.add(get_collection_id_from_cache_key(cache_key)[1])
+
+        # Generate collection elements that where is the cache.
+        for cache_key, cached_full_data in cached_full_data_dict.items():
+            yield CollectionElement.from_values(
+                *get_collection_id_from_cache_key(cache_key),
+                full_data=cached_full_data)
+
+        # Generate collection element that where not in the cache.
+        # If config elements are not in the cache, they have to be read from
```

Comment:
I do not understand this comment. How can we read configs from cache if they are not in? Do you mean 'config object' at the end of the sentence?